kernel/src/fs/sysfs/inode.rs

```diff
@@ -0,0 +1,617 @@
+        Ok(self.metadata.gid)
+    }
+    fn set_group(&self, _gid: Gid) -> Result<()> {
+        Err(Error::new(Errno::EPERM))
+    }
+
+    fn fs(&self) -> Arc<dyn FileSystem> {
+        crate::fs::sysfs::singleton().clone()
+    }
+
+    fn page_cache(&self) -> Option<crate::vm::vmo::Vmo<aster_rights::Full>> {
+        None
+    }
+
+    fn read_at(&self, offset: usize, buf: &mut VmWriter) -> Result<usize> {
+        self.read_direct_at(offset, buf)
+    }
+
+    fn read_direct_at(&self, _offset: usize, buf: &mut VmWriter) -> Result<usize> {
+        // TODO: it is unclear whether we should simply ignore the offset
+        // or report errors if it is non-zero.
+
+        let InnerNode::Attr(attr, leaf) = &self.inner_node else {
+            return Err(Error::new(Errno::EINVAL));
+        };
+
+        // TODO: check read permission
+
+        Err(Error::new(Errno::EINVAL))
+    }
+
+    fn write_at(&self, offset: usize, buf: &mut VmReader) -> Result<usize> {
+        self.write_direct_at(offset, buf)
+    }
+
+    fn write_direct_at(&self, _offset: usize, buf: &mut VmReader) -> Result<usize> {
+        let InnerNode::Attr(attr, leaf) = &self.inner_node else {
+            return Err(Error::new(Errno::EINVAL));
+        };
+
+        // TODO: check write permission
+
+        Err(Error::new(Errno::EINVAL))
+    }
+
+    fn create(&self, _name: &str, _type_: InodeType, _mode: InodeMode) -> Result<Arc<dyn Inode>> {
+        Err(Error::new(Errno::EPERM))
+    }
+
+    fn mknod(&self, _name: &str, _mode: InodeMode, _dev: MknodType) -> Result<Arc<dyn Inode>> {
+        Err(Error::new(Errno::EPERM))
+    }
+
+    fn link(&self, _old: &Arc<dyn Inode>, _name: &str) -> Result<()> {
+        Err(Error::new(Errno::EPERM))
+    }
+
+    fn unlink(&self, _name: &str) -> Result<()> {
+        Err(Error::new(Errno::EPERM))
+    }
+
+    fn rename(&self, _old_name: &str, _target: &Arc<dyn Inode>, _new_name: &str) -> Result<()> {
+        Err(Error::new(Errno::EPERM))
+    }
+
+    fn lookup(&self, name: &str) -> Result<Arc<dyn Inode>> {
+        if self.type_() != InodeType::Dir {
+            return Err(Error::new(Errno::ENOTDIR));
+        }
+
+        if name == "." {
+            return Ok(self.this());
+        } else if name == ".." {
+            return Ok(self.parent.upgrade().unwrap_or_else(|| self.this()));
+        }
+        match &self.inner_node {
+            InnerNode::Branch(obj) => {
+                if let Some(branch) = obj.as_branch() {
+                    self.lookup_node_or_attr(name, branch)
+                } else if let Some(node) = obj.as_node() {
+                    self.lookup_attr(name, node)
+                } else {
+                    Err(Error::new(Errno::ENOTDIR))
+                }
+            }
+            _ => Err(Error::new(Errno::ENOTDIR)),
+        }
+    }
+
+    /// Reads directory entries starting from a given offset.
+    ///
+    /// Why interpreting the `offset` argument as an inode number?
```

Comment:
Move the explanation about "offset" into the method, before a `min_ino` variable. These comments are for implementation, not the interface.